### ISSUE TITLE
GRPO rewards are not grouped as expected

### ISSUE THREAD

**Issue by erikchwang** [huggingface/trl#5030]
### Reproduction

https://github.com/huggingface/trl/blob/7c4e7f86047b82ad0e5ff7c8e3bb280b73024f31/trl/trainer/grpo_trainer.py#L1725-L1731

suppose there are 2 GPUs (G), 3 prompts (P), and 4 reward functions (F), 

`rewards_per_func` has been gathered from all GPUs and will be:

```
[
[G1_P1_F1, G1_P1_F2, G1_P1_F3, G1_P1_F4],
[G1_P2_F1, G1_P2_F2, G1_P2_F3, G1_P2_F4],
[G1_P3_F1, G1_P3_F2, G1_P3_F3, G1_P3_F4],
[G2_P1_F1, G2_P1_F2, G2_P1_F3, G2_P1_F4],
[G2_P2_F1, G2_P2_F2, G2_P2_F3, G2_P2_F4],
[G2_P3_F1, G2_P3_F2, G2_P3_F3, G2_P3_F4],
]
```

`rewards` will be:

```
[
G1_P1, 
G1_P2, 
G1_P3, 
G2_P1, 
G2_P2, 
G2_P3
]
```

and `rewards.view(-1, num_generations)` will be:

```
[
[G1_P1, G1_P2],
[G1_P3, G2_P1],
[G2_P2, G2_P3]
]
```

then `mean_grouped_rewards = rewards.view(-1, num_generations).mean(dim=1)` is not expected since rewards are not averaged across GPUs

**Comment by qgallouedec** [comment]
what does `G` stands for? GPU or generation?

**Comment by qgallouedec** [comment]
No, the GPU dispatching/aggregation actually works differently: inner variation is the generation, and outer is the prompt.
A simple script to show you that we group correctly:
Here we define a dummy reward function that gives +1 reward for all completions from prompt 1, and 0.0 for all completions from prompt 2.

```python
# accelerate launch --num_processes 2 5030.py
from datasets import load_dataset
from trl import GRPOConfig, GRPOTrainer

dataset = load_dataset("trl-internal-testing/zen", "standard_prompt_only", split="train")

def reward_func(prompts, completions, **kwargs):
    rewards=    []
    for prompt, completion in zip(prompts, completions):
        if "Although" in prompt: # first sample in the dataset
            rewards.append(1.0)
        else:
            rewards.append(0.0)
    return rewards

training_args = GRPOConfig(
    output_dir="tmp_dir",
    per_device_train_batch_size=3,  # reduce the batch size to reduce memory usage
    num_generations=3,  # reduce the number of generations to reduce memory usage
    max_completion_length=8,  # reduce the completion length to reduce memory usage
    report_to="none",
)
trainer = GRPOTrainer(
    model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
    reward_funcs=[reward_func],
    args=training_args,
    train_dataset=dataset,
)

trainer.train()
```

```python
rewards_per_func = self._calculate_rewards(inputs, prompts, completions, completion_ids_list)
# tensor([[1.],
#         [1.],
#         [1.],
#         [0.],
#         [0.],
#         [0.]], device='cuda:0')
```

```python
rewards = (rewards_per_func * self.reward_weights.to(device).unsqueeze(0)).nansum(dim=1)
# tensor([1., 1., 1., 0., 0., 0.], device='cuda:0')
```

```python
mean_grouped_rewards = rewards.view(-1, num_generations).mean(dim=1)
# tensor([1., 0.], device='cuda:0')
```

**Comment by erikchwang** [comment]
> what does `G` stands for? GPU or generation?

G means GPUs. P means promps. F means reward functions.

GRPOTrainer assume num_generation is the same as num_gpus

**Comment by qgallouedec** [comment]
The same works with `num_generations=2` (not recommended) and 2 GPUs (per_device_batch_size=3, hence num_generations*per_device_batch_size/num_gpu=3 groups per iteration):

```python
>>> rewards_per_func
tensor([[1.],
        [1.],
        [0.],
        [0.],
        [0.],
        [0.]], device='cuda:0')
>>> rewards
tensor([1., 1., 0., 0., 0., 0.], device='cuda:0')
>>> mean_grouped_rewards
tensor([1., 0., 0.], device='cuda:0')
```

**Comment by erikchwang** [comment]
your num_generations should be exactly the same as num_gpus?

**Comment by qgallouedec** [comment]
We never  "averaged between GPUs" because the prompt-completion distribution depends on the GAS, number of completions, and per device batch size. Can you give a MRE in the spirit of https://github.com/huggingface/trl/issues/5030#issuecomment-3872681941 to show your point?

**Comment by qgallouedec** [comment]
- We never  "average between GPUs" because the prompt-completion distribution depends on the GAS, number of completions, and per device batch size. Can you give a MRE in the spirit of https://github.com/huggingface/trl/issues/5030#issuecomment-3872681941 to show your point?

**Comment by erikchwang** [comment]
https://github.com/huggingface/trl/blob/db0d95523e5b8039c94c50df1b6286ff8b7e29ce/trl/trainer/grpo_trainer.py#L830

According to `_get_train_sampler`, each prompt has 1 generation on each GPU

https://github.com/huggingface/trl/blob/db0d95523e5b8039c94c50df1b6286ff8b7e29ce/trl/trainer/grpo_trainer.py#L1152

According to `_calculate_rewards`, `rewards_per_func` is gathered from all GPUs, so it will be:

```
 [
[G1_P1_F1, ...],
[G1_P2_F1, ...],
[G1_P3_F1, ...],
[G2_P1_F1, ...],
[G2_P2_F1, ...],
[G2_P3_F1, ...],
]
```
Then, `rewards = (rewards_per_func * self.reward_weights.to(device).unsqueeze(0)).nansum(dim=1)` will be:

```
[
G1_P1, 
G1_P2, 
G1_P3, 
G2_P1, 
G2_P2, 
G2_P3
]
```

Then, `rewards.view(-1, num_generations)` will be:

```
[
[G1_P1, G1_P2],
[G1_P3, G2_P1],
[G2_P2, G2_P3]
]
```

But this is wrong, because `rewards.view(-1, num_generations).mean(dim=1)` is taken on dim=1, like between [G1_P1, G1_P2], but this is not an expected group!

The expected groups should be:

```
[
[G1_P1, G2_P1],
[G1_P2, G2_P2],
[G1_P3, G2_P3],
]
```

**Comment by erikchwang** [comment]
This bug does not break the running, but rewards are not grouped as expected, this will affect performance

**Comment by erikchwang** [comment]
> * We never  "average between GPUs" because the prompt-completion distribution depends on the GAS, number of completions, and per device batch size. Can you give a MRE in the spirit of [GRPO rewards are not grouped as expected #5030 (comment)](https://github.com/huggingface/trl/issues/5030#issuecomment-3872681941) to show your point?

I have run your code in debug mode, the bug really exists

**Comment by erikchwang** [comment]
I made a small change to the reward function to make the bug clear: 

```
# accelerate launch --num_processes 2 test.py
from datasets import load_dataset
from trl import GRPOConfig, GRPOTrainer

dataset = load_dataset(
    "trl-internal-testing/zen", "standard_prompt_only", split="train"
)


def reward_func(prompts, completions, **kwargs):
    rewards = [float(i) for i in range(len(prompts))]
    return rewards


training_args = GRPOConfig(
    output_dir="tmp_dir",
    per_device_train_batch_size=3,  # reduce the batch size to reduce memory usage
    num_generations=2,  # reduce the number of generations to reduce memory usage
    max_completion_length=8,  # reduce the completion length to reduce memory usage
    report_to="none",
)
trainer = GRPOTrainer(
    model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
    reward_funcs=[reward_func],
    args=training_args,
    train_dataset=dataset,
)

trainer.train()
```

Using 2 GPUs to debug: 

before `rewards_per_func = gather(rewards_per_func)`, you will get `rewards_per_func = [[0],[1],[2]]` on each GPU

after `rewards_per_func = gather(rewards_per_func)`, you will get `rewards_per_func = [[0],[1],[2],[0],[1],[2]]` (**this is where the error starts, due to the behaviour of `gather`**)

then you will get `rewards = [0., 1., 2., 0., 1., 2.]`

and `rewards.view(-1, num_generations)` will be [[0, 1], [2, 0], [1, 2]], these groups are wrong! 

The correct groups should be [[0, 0], [1, 1], [2, 2]]

**Comment by qgallouedec** [comment]
No, 
- GPU 0 will generate 2 completions for prompt 0 and 1 for prompt 1
- GPU 1 will generate 1 completion for prompt 1 and 2 for prompt 2

See https://github.com/huggingface/trl/blob/db0d95523e5b8039c94c50df1b6286ff8b7e29ce/trl/trainer/grpo_trainer.py#L846

**Comment by erikchwang** [comment]
OK I see... I misunderstood the distribution of prompts...

I used a dummy dataset to debug: 

```
# accelerate launch --num_processes 2 test.py
from datasets import Dataset
from trl import GRPOConfig, GRPOTrainer

dataset = Dataset.from_dict(
    {
        "prompt": [str(float(i)) for i in range(100)],
    }
)


def reward_func(prompts, completions, **kwargs):
    rewards = [float(i) for i in prompts]
    return rewards


training_args = GRPOConfig(
    output_dir="tmp_dir",
    per_device_train_batch_size=3,  # reduce the batch size to reduce memory usage
    num_generations=2,  # reduce the number of generations to reduce memory usage
    max_completion_length=8,  # reduce the completion length to reduce memory usage
    report_to="none",
)
trainer = GRPOTrainer(
    model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
    reward_funcs=[reward_func],
    args=training_args,
    train_dataset=dataset,
)

trainer.train()
```
everything is expected, thanks!

**Comment by qgallouedec** [comment]
In your example, the reward function 
- in process 0 will get completion P0C0 (prompt 0 completion 0), P0C1 and P1C0
- in process 1 will get completion P1C1, P2C0 and P2C1

So

- P0C0: reward=0
- P0C1: reward=1
- P1C0: reward=2
- P1C1: reward=0
- P2C0: reward=1
- P2C1: reward=2

Thus, per group:

- P0: [0, 1]
- P1: [2, 0]
- P2: [1, 2]

This matches what you get

**Comment by qgallouedec** [comment]
Maybe the confusion comes from the fact that the reward function in called per process, not per group or per completion idx
